### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by ericlplanteschool**
implemented base requirements and unit tests

reviewed @MikePHaynes

### CODE REVIEW

**Comment by MikePHaynes** [comment]
Will review this!

**Review by MikePHaynes — COMMENTED** [review]
Very good overall, it looks like you covered all of the requirements besides the first one. I could be mistaken, but it doesn't look like the ClassName auto property is set in LogFactory with an object initializer. Also, I believe that the purpose of the ConfigureFileLogger method in the context of the CreateLogger method is to make sure that a file path exists so that if one doesn't then you can return null in CreateLogger, although I think technically you covered this requirement. Finally, Visual Studio came up with quite a few messages regarding your code but fortunately I believe they all should be very quick & easy fixes. I'll highlight the particular lines of code below this comment.

## Assignment

- There is an existing `BaseLogger` class. It needs an **auto property** to hold the class name. This property should be set in the `LogFactory` using an **object initializer**. ❌
- Create a `FileLogger` that derives from `BaseLogger`. It should take in a path to a file to write the log message to. When its `Log` method is called, it should **append** messages on their own line in the file. The output should include all of the following:
  - The current date/time ✔
  - The name of the class that created the logger ✔
  - The log level ✔
  - The message ✔
  - The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
- The `LogFactory` should be updated with a new method `ConfigureFileLogger`. This should take in a file path and store it in a **private member**. It should use this when instantiating a new `FileLogger` in its `CreateLogger` method. ✔
- If the file logger has not be configured in the `LogFactory`, its `CreateLogger` method should return `null`. ✔
- Inside of `BaseLoggerMixins` implement **extension methods** on `BaseLogger` for
  - `Error`, ✔
  - `Warning`, ✔
  - `Information`, and ✔
  - `Debug`. ✔
  Each of these methods should take in a `string` for the message, as well as a **parameter array** of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the `BaseLogger.Log` method, by automatically supplying the appropriate `LogLevel`. These methods should throw an exception if the `BaseLogger` parameter is null. There are a couple example unit tests to get you started.
- Use the nameof() operator when identifying the class name to the logger ✔
- Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
- Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
- Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
- Set `LangVersion` to 10.0 ✔
- Turn on Nullability (`Nullable`) ✔
- **All of the above should be unit tested.**

## Extra Credit

- Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses `System.Console` or `System.Diagnostics.Trace` ❌

File: Logger/BaseLogger.cs [hunk 1]
```diff
@@ -2,6 +2,7 @@
 {
     public abstract class BaseLogger
     {
+        public string className { get; set; }
```
Comment (MikePHaynes) [inline Logger/BaseLogger.cs:5]: Visual Studio says this should be ClassName.

File: Logger/BaseLoggerMixins.cs [hunk 2]
```diff
@@ -1,7 +1,52 @@
-﻿namespace Logger
+﻿using System;
+
+namespace Logger.Mixins
 {
     public static class BaseLoggerMixins
     {
+        public static void errorLog(this BaseLogger logger, string message, params string[] arr)
```
Comment (MikePHaynes) [inline Logger/BaseLoggerMixins.cs:7]: Visual Studio says this should say ErrorLog

File: Logger/BaseLoggerMixins.cs [hunk 3]
```diff
@@ -1,7 +1,52 @@
...
+            if(logger == null){
+                throw new ArgumentNullException();
+            }
+        
+            string res = String.Format(message, arr);
+        
+            logger.Log(LogLevel.Error, res);
+        }
+
+        public static void warningLog(this BaseLogger logger, string message, params string[] arr)
```
Comment (MikePHaynes) [inline Logger/BaseLoggerMixins.cs:18]: Visual Studio says this should say WarningLog

File: Logger/BaseLoggerMixins.cs [hunk 4]
```diff
@@ -1,7 +1,52 @@
...
+            if(logger == null){
+                throw new ArgumentNullException();
+            }
+
+            string res = String.Format(message, arr);
+
+            logger.Log(LogLevel.Warning, res);
+        }
+
+        public static void informationLog(this BaseLogger logger, string message, params string[] arr)
```
Comment (MikePHaynes) [inline Logger/BaseLoggerMixins.cs:29]: Visual Studio says this should say InformationLog

File: Logger/BaseLoggerMixins.cs [hunk 5]
```diff
@@ -1,7 +1,52 @@
...
+            if(logger == null){
+                throw new ArgumentNullException();
+            }
+        
+            string res = String.Format(message, arr);
+
+            logger.Log(LogLevel.Information, res);
+        }
+
+        public static void debugLog(this BaseLogger logger, string message, params string[] arr)
```
Comment (MikePHaynes) [inline Logger/BaseLoggerMixins.cs:40]: Visual Studio says this should say DebugLog

File: Logger/FileLogger.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
+using System;
+using System.IO;
+using Logger.Mixins;
+
+namespace Logger{
+
+    public class FileLogger : BaseLogger
+    {
+        public string filePath { get; set; }
```
Comment (MikePHaynes) [inline Logger/FileLogger.cs:9]: Visual Studio says this should be FilePath

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 7]
```diff
@@ -26,14 +29,59 @@ public void Error_WithData_LogsMessage()
...
+            logger.errorLog("Message {0}", "42");
 
             // Assert
             Assert.AreEqual(1, logger.LoggedMessages.Count);
             Assert.AreEqual(LogLevel.Error, logger.LoggedMessages[0].LogLevel);
             Assert.AreEqual("Message 42", logger.LoggedMessages[0].Message);
         }
 
+        [TestMethod]
+        public void warning_WithData_LogsMessage()
```
Comment (MikePHaynes) [inline Logger.Tests/BaseLoggerMixinsTests.cs:41]: Use PascalCase with format NameOfMethodYourTesting_Condition_ExpectedResult

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 8]
```diff
@@ -26,14 +29,59 @@ public void Error_WithData_LogsMessage()
...
+            logger.warningLog("Message {0}", "42");
+
+            // Assert
+            Assert.AreEqual(1, logger.LoggedMessages.Count);
+            Assert.AreEqual(LogLevel.Warning, logger.LoggedMessages[0].LogLevel);
+            Assert.AreEqual("Message 42", logger.LoggedMessages[0].Message);
+        }
+
+        [TestMethod]
+        public void information_WithData_LogsMessage()
```
Comment (MikePHaynes) [inline Logger.Tests/BaseLoggerMixinsTests.cs:56]: Use PascalCase with format NameOfMethodYourTesting_Condition_ExpectedResult

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 9]
```diff
@@ -26,14 +29,59 @@ public void Error_WithData_LogsMessage()
...
+            logger.informationLog("Message {0}", "42");
+
+            // Assert
+            Assert.AreEqual(1, logger.LoggedMessages.Count);
+            Assert.AreEqual(LogLevel.Information, logger.LoggedMessages[0].LogLevel);
+            Assert.AreEqual("Message 42", logger.LoggedMessages[0].Message);
+        }
+
+        [TestMethod]
+        public void debug_WithData_LogsMessage()
```
Comment (MikePHaynes) [inline Logger.Tests/BaseLoggerMixinsTests.cs:71]: Use PascalCase with format NameOfMethodYourTesting_Condition_ExpectedResult

File: Logger.Tests/FileLoggerTests.cs [hunk 10]
```diff
@@ -1,10 +1,29 @@
...
+using Logger.Mixins;
+using System;
 
 namespace Logger.Tests
 {
     [TestClass]
     public class FileLoggerTests
     {
+        [TestMethod]
+        public void LogTest()
```
Comment (MikePHaynes) [inline Logger.Tests/FileLoggerTests.cs:11]: Use PascalCase with format NameOfMethodYourTesting_Condition_ExpectedResult

File: Logger.Tests/LogFactoryTests.cs [hunk 11]
```diff
@@ -6,6 +6,36 @@ namespace Logger.Tests
     [TestClass]
     public class LogFactoryTests
     {
+        [TestMethod]
+        public void createlogger_returns_null_when_not_configured()
```
Comment (MikePHaynes) [inline Logger.Tests/LogFactoryTests.cs:10]: Use PascalCase with format NameOfMethodYourTesting_Condition_ExpectedResult

File: Logger.Tests/LogFactoryTests.cs [hunk 12]
```diff
@@ -6,6 +6,36 @@ namespace Logger.Tests
...
+            
+            // Act
+            FileLogger res = (FileLogger) factory.CreateLogger("","doesntexist");
+            
+            // Assert
+            Assert.AreEqual(null, res);
+        }
+
+        [TestMethod]
+        public void factory_creates_logger()
```
Comment (MikePHaynes) [inline Logger.Tests/LogFactoryTests.cs:23]: Use PascalCase with format NameOfMethodYourTesting_Condition_ExpectedResult

**Comment by MikePHaynes** [comment]
Final note: I recall Mark mentioning in class that we should avoid using "this.Whatever" whenever possible in our code ("this BaseLogger logger" is perfectly fine to be clear) and double check the messages in the "Error List" in visual studio. Again, most if not all of the fixes will just take a second or two to fix. Otherwise terrific work!

File: Logger/FileLogger.cs [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+        
+        
+        public override void Log(LogLevel logLevel, string message)
+        {
+            StreamWriter writer = new StreamWriter(filePath);
+
+            writer.WriteLine(DateTime.Now.ToString() + " " + nameof(this.className)
+            + " " + logLevel + " " + message);
+
+            writer.Close();
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:19]: You could also use [`using`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement)

File: Logger/FileLogger.cs [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    public class FileLogger : BaseLogger
+    {
+        public string filePath { get; set; }
+        
+        
+        public override void Log(LogLevel logLevel, string message)
+        {
+            StreamWriter writer = new StreamWriter(filePath);
+
+            writer.WriteLine(DateTime.Now.ToString() + " " + nameof(this.className)
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:16]: You shouldn't need to use `nameOf` here, that should be used when setting the class name. Also, avoid using `this`

File: Logger/LogFactory.cs [hunk 15]
```diff
@@ -1,11 +1,43 @@
...
-            return null;
+        }
+
+        public BaseLogger CreateLogger(string className, string path)
+        {
+            this.ConfigureFileLogger(path);
+
+            if(this.configured == false){
+                return null;
+            }
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:31]: Not quite what was asked - you should be checking if the path is null or not here.

File: Logger/LogFactory.cs [hunk 16]
```diff
@@ -1,11 +1,43 @@
...
+
+        private string path;
+
+        public void ConfigureFileLogger(string path)
         {
+            if(File.Exists(path))
+            {
+                this.path = path;
+
+                this.configured = true;
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:18]: While this will set configured to false if the path is null, it's not quite what was asked of you in the assignment as it will also be set to false if the path doesn't exist, which may not be desirable in all cases.

File: Logger.Tests/LogFactoryTests.cs [hunk 17]
```diff
@@ -6,6 +6,36 @@ namespace Logger.Tests
...
+        {
+            // Arrange
+            LogFactory factory = new LogFactory();
+            
+            // Act
+            FileLogger res = (FileLogger) factory.CreateLogger("class","testfile.txt");
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:29]: You may consider storing theses strings in a variable so you don't type them out twice.

File: Logger.Tests/LogFactoryTests.cs (hunk 11 above)
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:10]: You may also want to test this with null input.

File: Logger.Tests/FileLoggerTests.cs (hunk 10 above)
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:11]: Good job, other than what MikePHaynes mentioned.

**Review by mmwoodfo — COMMENTED** [review]
There are a lot of good comments here about naming guidelines. One of the things you are graded on is if you fix issues pointed out in a review done by someone else.
